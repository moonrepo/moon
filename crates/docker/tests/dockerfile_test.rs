use moon_common::Id;
use moon_docker::{GenerateDockerfileOptions, generate_dockerfile};
use moon_target::Target;
use starbase_sandbox::assert_snapshot;

fn create_options() -> GenerateDockerfileOptions {
    GenerateDockerfileOptions {
        project: Id::raw("app"),
        ..Default::default()
    }
}

mod dockerfile {
    use super::*;

    #[test]
    fn renders_defaults() {
        assert_snapshot!(generate_dockerfile(create_options()).unwrap());
    }

    #[test]
    fn disables_toolchain() {
        let mut options = create_options();
        options.disable_toolchain = true;

        assert_snapshot!(generate_dockerfile(options).unwrap());
    }

    #[test]
    fn with_tasks() {
        let mut options = create_options();
        options.build_task = Some(Target::parse("app:compile").unwrap());
        options.start_task = Some(Target::parse("app:serve").unwrap());

        assert_snapshot!(generate_dockerfile(options).unwrap());
    }

    #[test]
    fn with_prune() {
        let mut options = create_options();
        options.prune = true;
        options.build_task = Some(Target::parse("app:compile").unwrap());
        options.start_task = Some(Target::parse("app:serve").unwrap());

        assert_snapshot!(generate_dockerfile(options).unwrap());
    }
}
