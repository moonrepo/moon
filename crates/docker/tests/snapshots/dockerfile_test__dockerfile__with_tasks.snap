---
source: crates/docker/tests/dockerfile_test.rs
expression: generate_dockerfile(options).unwrap()
---
#### BASE STAGE
#### Installs moon.

FROM scratch AS base
WORKDIR /app

# Install moon binary
RUN curl -fsSL https://moonrepo.dev/install/moon.sh | bash
ENV PATH="/root/.moon/bin:/root/.proto/bin:$PATH"

#### SKELETON STAGE
#### Scaffolds repository skeleton structures.

FROM base AS skeleton

# Copy entire repository and scaffold
COPY . .
RUN moon docker scaffold app

#### BUILD STAGE
#### Builds the project.

FROM base AS build

# Copy workspace configs
COPY --from=skeleton /app/.moon/docker/workspace .

# Install dependencies
RUN moon docker setup

# Copy project sources
COPY --from=skeleton /app/.moon/docker/sources .

# Build the project
RUN moon run app:compile

#### START STAGE
#### Runs the project.

FROM base AS start

# Copy built sources
COPY --from=build /root/.proto /root/.proto
COPY --from=build /app /app

CMD moon run app:serve
